src/main/scala/com/fulcrumgenomics/rnaseq/AnnotateByIntervalList.scala

```diff
@@ -0,0 +1,246 @@
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(this.input)
+  Io.assertReadable(this.intervals)
+  Io.assertCanWriteFile(this.output)
+
+  override def execute(): Unit = {
+    val progress = ProgressLogger(logger, verb = "Processed", noun = "loci", unit = 1000000)
+    val reader = SamSource(this.input)
+    val dict = reader.dict
+
+    val (baseCounter, uniqueBaseCounter, templateCounter, uniqueTemplateCounter, overlapDetector) = {
+      val _baseCounter            = new SimpleCounter[String]()
+      val _uniqueBaseCounter      = new SimpleCounter[String]()
+      val _templateCounter        = new SimpleCounter[String]()
+      val _uniqueTemplateCounter  = new SimpleCounter[String]()
+      val intervals               = buildIntervals(dict.asSam)
+      intervals.foreach { interval => _baseCounter.count(interval.getName, 0) }
+
+      val _overlapDetector = OverlapDetector.create[Interval](intervals.iterator.toJavaList)
+      (_baseCounter, _uniqueBaseCounter, _templateCounter, _uniqueTemplateCounter, _overlapDetector)
+    }
+
+    // Templates are assigned to an interval together.
+    val iterator = reader.iterator
+    val records = iterator.filterNot(record => record.duplicate
+      || record.secondary
+      || record.supplementary
+      || !record.pf
+      || !record.isFrPair
+      || record.mapq < minMappingQuality)
+
+    // First we filter to ensure that we only pick one read in a template and then iterate over those reads to create
+    // an interval for each template. For single ended reads the interval is just the read.
+    val templateIntervals = records.filter(record => {
+      if (record.mateMapped) {
+        val equalStartTieBreaker =
+          if (Math.min(record.start, record.end) == Math.min(record.mateStart, record.mateEnd.getOrElse(Int.MaxValue))) {
+            record.firstOfPair
+          } else {
+            true
+          }
+        // Take only the first read by coordinate order
+        Math.min(record.start, record.end) < Math.min(record.mateStart, record.mateEnd.getOrElse(Int.MaxValue)) &&
+        // if mate start == start pick first of pair
+        equalStartTieBreaker
+      } else {
+        //If we are single ended then just take the record
+        true
+      }
+    }).map(record =>
+      // Interval should be end of read one to the beginning of read two
+      new Interval(
+        record.refName,
+        Math.max(record.start, record.end) + 1,
+        Math.min(record.mateStart, record.mateEnd.getOrElse(Int.MaxValue)) - 1
+      )
+    )
+
+    var numLociMultiMatched: Long = 0
+
+    templateIntervals.foreach { templateInterval =>
+      // Overlaps should only be flagged when they enclose the template bases.
+      val overlaps = overlapDetector.getOverlaps(templateInterval).toSeq
+        .filter(interval => CoordMath.encloses(interval.getStart, interval.getEnd,
+          templateInterval.getStart, templateInterval.getEnd))
```

Comment:
My preference is a single line for a function call, or named parameters on multiple lines